kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+                    self.state.board.lichen[unit.pos.y, unit.pos.x] = max(self.state.board.lichen[unit.pos.y, unit.pos.x] - unit.unit_cfg.DIG_LICHEN_REMOVED, 0)
+                elif self.state.board.ice[unit.pos.y, unit.pos.x] > 0:
+                    unit.add_resource(0, unit.unit_cfg.DIG_RESOURCE_GAIN)
+                elif self.state.board.ore[unit.pos.y, unit.pos.x] > 0:
+                    unit.add_resource(1, unit.unit_cfg.DIG_RESOURCE_GAIN)
+                unit.power -= math.ceil(self.state.env_cfg.ROBOTS[unit.unit_type.name].DIG_COST * weather_cfg["power_loss_factor"])
+
+            for unit, self_destruct_action in actions_by_type["self_destruct"]:
+                unit: Unit
+                self_destruct_action: SelfDestructAction
+                pos_hash = self.state.board.pos_hash(unit.pos)
+                del self.state.board.units_map[pos_hash]
+                self.destroy_unit(unit)
+
+            for factory, factory_build_action in actions_by_type["factory_build"]:
+                factory: Factory
+                factory_build_action: FactoryBuildAction
+                team = self.state.teams[factory.team.agent]
+                self.add_unit(
+                    team=team,
+                    unit_type=factory_build_action.unit_type,
+                    pos=factory.pos.pos,
+                )
+                if factory_build_action.unit_type == UnitType.HEAVY:
+                    factory.sub_resource(3, self.env_cfg.ROBOTS["HEAVY"].METAL_COST)
+                    factory.sub_resource(4, math.ceil(self.env_cfg.ROBOTS["HEAVY"].POWER_COST * weather_cfg["power_loss_factor"]))
+                else:
+                    factory.sub_resource(3, self.env_cfg.ROBOTS["LIGHT"].METAL_COST)
+                    factory.sub_resource(4, math.ceil(self.env_cfg.ROBOTS["LIGHT"].POWER_COST * weather_cfg["power_loss_factor"]))
+
+            new_units_map: Dict[str, List[Unit]] = defaultdict(list)
+            heavy_entered_pos: Dict[str, List[Unit]] = defaultdict(list)
+            light_entered_pos: Dict[str, List[Unit]] = defaultdict(list)
+
+            for unit, move_action in actions_by_type["move"]:
+                move_action: MoveAction
+                # skip move center
+                if move_action.move_dir == 0:
+                    continue
+                old_pos_hash = self.state.board.pos_hash(unit.pos)
+                target_pos = unit.pos + move_action.dist * move_deltas[move_action.move_dir]
+                rubble = self.state.board.rubble[target_pos.y, target_pos.x]
+                power_required = unit.unit_cfg.MOVE_COST + unit.unit_cfg.RUBBLE_MOVEMENT_COST * rubble
+                power_required = math.ceil(power_required * weather_cfg["power_loss_factor"])
+                unit.pos = target_pos
+                new_pos_hash = self.state.board.pos_hash(unit.pos)
+
+                # remove unit from map temporarily
+                if len(self.state.board.units_map[old_pos_hash]) == 1:
```

Comment:
shouldn't we only do this if they have enough power?

Reply:
I assume you mean move a unit if they have enough power. An earlier function removes all invalid actions including those where the unit doesn't have enough power. So these actions should all be valid on their own.